### DISCUSSION TITLE
Convert Cosine Similarity to Probability

### DISCUSSION THREAD

**Discussion by gor2000** [NVIDIA-NeMo/NeMo#10430] [Q&A]
Hi I'm trying to understand in this model how should I convert in some way the cosine similarity to a probability.

I'm working for my company in a project where we need to detect if 2 audios have the same speaker or not. Basically, trying to detect if an audio has been cloned or it's original. 

The client has required the output of the application that detects audios to be a percentage. That is, 100% should be max similarity and 0% it's not the same speaker at all. 

My question is: Is it possible to do that?. Does it make sense? Because what I understant is that cosine similarity = 0 does not mean 50% right?

I would really appreciate any help here as I have not found anything on the internet.

Thank you in advance!!

**Comment by zhenyih** [comment]
A response derived from NEMO source and Agentic AI.   Please verify accuracy.

To convert cosine similarity scores from a speaker verification model to a percentage-based probability (0-100%), you can use several approaches:

1. Linear scaling (simplest): Convert the cosine similarity from [-1, 1] range to [0, 1] using:
   ```python
   probability = (similarity + 1) / 2 * 100
   ```
   This is already implemented in NeMo's speaker verification code.

2. Sigmoid transformation (better probability calibration):
   ```python
   import numpy as np
   def similarity_to_probability(similarity, alpha=10, beta=0.5):
       return 100 / (1 + np.exp(-alpha * (similarity - beta)))
   ```
   You'll need to tune alpha (controls steepness) and beta (threshold) on your validation data.

3. Logistic regression calibration (most accurate):
   Train a logistic regression model on labeled pairs of embeddings to predict probability from similarity scores.

For your application detecting if audio has been cloned, option 2 (sigmoid) likely offers the best balance between simplicity and accuracy. The parameters can be adjusted based on your specific use case - increase beta to make the system more conservative about declaring matches.

Remember that any percentage threshold you choose should be validated on your specific dataset to ensure it meets your requirements for false positives and false negatives.
